### PR TITLE
Tk/row details

### PR DESCRIPTION
**Pull request by TKonov**

### CODE REVIEW

File: Controls/Grid/Grid.UWP/Model/TableModel/DisplayData/CellsController.SlotLength.cs [hunk 1]
```diff
@@ -122,14 +126,14 @@ internal virtual double GetSlotHeight(int cellSlot)
...
         internal bool UpdateSlotHeight(int cellSlot, double cellHeight)
         {
-            if (this.table.RowHeightIsNaN && !this.RowPool.IsItemCollapsed(cellSlot))
+            if (this.table.RowHeightIsNaN && !this.RowPool.IsItemCollapsed(cellSlot) || this.table.HasExpandedRowDetails(cellSlot))
             {
                 var renderInfo = this.RowPool.RenderInfo;
                 var currentHeight = renderInfo.ValueForIndex(cellSlot, false);
 
-                bool isLessThan = GridModel.DoubleArithmetics.IsLessThan(currentHeight, cellHeight);
+                bool isLessThan = true;
```
Comment (APopatanasov) [inline Controls/Grid/Grid.UWP/Model/TableModel/DisplayData/CellsController.SlotLength.cs]: This is not changed anywhere. We could simply return true in the if statement after it.

File: Controls/Grid/Grid.UWP/Model/TableModel/DisplayData/CellsController.cs [hunk 2]
```diff
@@ -76,18 +76,20 @@ internal RadSize OnCellsArrange(RadSize finalSize)
...
+                    int line = rowModel.ItemInfo.Slot;
+                    topOffset = line - 1 >= 0 ? (this.RowPool.RenderInfo.OffsetFromIndex(line - 1) - rowModel.RowDetailsSize.Height) : 0;
                 }
 
                 Dictionary<int, T> columnCellsPair;
                 this.generatedRowCells.TryGetValue(rowSlot, out columnCellsPair);
 
                 leftOffset = 0;
-                double cellHeight = this.GetSlotHeight(rowSlot);
+                double rowSlotHeight = this.GetSlotHeight(rowSlot);
```
Comment (APopatanasov) [inline Controls/Grid/Grid.UWP/Model/TableModel/DisplayData/CellsController.cs]: Do we need both rowSlotHeight and cellHeight - it seems the cellHeight is equal to the rowSlotHeight? Shouldn't we use only one of them as the previous implementation?

File: Controls/Grid/Grid.UWP/View/Services/RowDetails/RowDetailsService.cs [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+        internal HashSet<object> ExpandedItems { get; private set; }
+
+        public RowDetailsService(RadDataGrid owner)
+            : base(owner)
+        {
+            this.ExpandedItems = new HashSet<object>();
+
+            this.DetailsPresenter = new DataGridRowDetailsControl();
+
```
Comment (APopatanasov) [inline Controls/Grid/Grid.UWP/View/Services/RowDetails/RowDetailsService.cs]: Additional empty row

File: Controls/Grid/Grid.UWP/View/Services/RowDetails/RowDetailsService.cs [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+                this.ExpandedItems.Add(item);
+                this.Owner.updateService.RegisterUpdate((int)UpdateFlags.AffectsContent);
+            }
+        }
+
+        public void CollapseDetailsForItem(object item)
+        {
+            this.ExpandedItems.Remove(item);
+            this.Owner.updateService.RegisterUpdate((int)UpdateFlags.AffectsContent);
+
```
Comment (APopatanasov) [inline Controls/Grid/Grid.UWP/View/Services/RowDetails/RowDetailsService.cs]: Delete row

File: Controls/Grid/Grid.UWP/View/Services/RowDetails/RowDetailsService.cs [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+            }
+
+            this.Owner.updateService.RegisterUpdate((int)UpdateFlags.AffectsContent);
+        }
+
+        public bool HasExpandedRowDetails(object item)
+        {
+            return this.ExpandedItems.Contains(item);
+        }
+
```
Comment (APopatanasov) [inline Controls/Grid/Grid.UWP/View/Services/RowDetails/RowDetailsService.cs]: Delete row

**Comment by JasperE84** [comment]
Very cool! Is there also a way to bind rowdetails visibility?
Now I'm calling a method on the control in the view from the viewmodel which isn't very mvvm-like.

```
this.MyGridControl.ItemSelected += (object sender, RpcObject e) => 
{
	if (this.NavigationService.Content is MyPage page)
	{
		page.MyGrid.ShowRowDetailsForItem(e);
	}
};
```

**Pull request merged**